File: app/src/main/java/social/entourage/android/user/edit/UserEditAboutFragment.java
package social.entourage.android.user.edit;


import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.core.content.res.ResourcesCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import social.entourage.android.R;
import social.entourage.android.api.model.User;
import social.entourage.android.user.UserFragment;

/**
 * Fragment to edit an user's about text
 */

public class UserEditAboutFragment extends DialogFragment {

    // ----------------------------------
    // CONSTANTS
    // ----------------------------------

    public static final String TAG = UserEditAboutFragment.class.getSimpleName();

    private static final int ABOUT_MAX_CHAR_COUNT = 200;

    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------

    @BindView(R.id.user_edit_about)
    EditText aboutEditText;

    @BindView(R.id.user_edit_about_char_count)
    TextView aboutCharCountTextView;

    private User user;
    private UserEditFragment userEditFragment;

    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------

    public UserEditAboutFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        if (getDialog() != null && getDialog().getWindow() != null) {
            getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
            getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);
        }
        View view = inflater.inflate(R.layout.fragment_user_edit_about, container, false);
        ButterKnife.bind(this, view);

        return view;
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        configureView();
    }

    @Override
    public void dismiss() {
        if (userEditFragment != null) {
            userEditFragment.scrollToOriginalPosition();
        }
        super.dismiss();
    }

    // ----------------------------------
    // BUTTONS HANDLING
    // ----------------------------------

    @OnClick(R.id.user_edit_about_close_button)
    protected void onCloseClicked() {
        dismiss();
    }

    @OnClick(R.id.user_edit_about_save_button)
    protected void onSaveClicked() {
        String about = aboutEditText.getText().toString();
        if (about.length() > ABOUT_MAX_CHAR_COUNT) {
            Toast.makeText(getContext(), R.string.user_edit_about_error, Toast.LENGTH_SHORT).show();
        } else {
            if (user != null) {
                user.setAbout(about.trim());
                if (userEditFragment != null) {
                    userEditFragment.configureView();
                } else {
                    if (getFragmentManager() != null) {
                        UserFragment userFragment = (UserFragment) getFragmentManager().findFragmentByTag(UserFragment.TAG);
                        if (userFragment != null) {
                            userFragment.saveAccount(user);
                        }
                    }
                }
            }
            dismiss();
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------

    private void configureView() {
        if (getFragmentManager() != null) {
            userEditFragment = (UserEditFragment) getFragmentManager().findFragmentByTag(UserEditFragment.TAG);
            if (userEditFragment == null) {
                //Try to see if we can find the User View fragment
                UserFragment userFragment = (UserFragment) getFragmentManager().findFragmentByTag(UserFragment.TAG);
                if (userFragment != null) {
                    user = userFragment.getEditedUser();
                }
            } else {
                user = userEditFragment.getEditedUser();
            }
        }
        if (user == null) {
            return;
        }

        aboutEditText.setText(user.getAbout());
        aboutEditText.setSelection(aboutEditText.length());
        aboutEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(final CharSequence s, final int start, final int count, final int after) {

            }

            @Override
            public void onTextChanged(final CharSequence s, final int start, final int before, final int count) {

            }

            @Override
            public void afterTextChanged(final Editable s) {
                String charCountString = getContext().getString(R.string.entourage_create_title_char_count_format, s.length(), ABOUT_MAX_CHAR_COUNT);
                aboutCharCountTextView.setText(charCountString);
                if (s.length() > ABOUT_MAX_CHAR_COUNT) {
                    aboutCharCountTextView.setTextColor(ResourcesCompat.getColor(getResources(), R.color.entourage_error, null));
                } else {
                    aboutCharCountTextView.setTextColor(ResourcesCompat.getColor(getResources(), R.color.entourage_ok, null));
                }
            }
        });

        String charCountString = getContext().getString(R.string.entourage_create_title_char_count_format, aboutEditText.length(), ABOUT_MAX_CHAR_COUNT);
        aboutCharCountTextView.setText(charCountString);
    }

}

File: app/src/main/java/social/entourage/android/user/edit/UserEditAboutFragment.kt
package social.entourage.android.user.edit

import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.DialogFragment
import kotlinx.android.synthetic.main.fragment_user_edit_about.*
import social.entourage.android.R
import social.entourage.android.api.model.User
import social.entourage.android.user.UserFragment

/**
 * Fragment to edit an user's about text
 */
class UserEditAboutFragment  : DialogFragment() {
    // ----------------------------------
    // ATTRIBUTES
    // ----------------------------------
    private var user: User? = null
    private var userEditFragment: UserEditFragment? = null
    // ----------------------------------
    // LIFECYCLE
    // ----------------------------------
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        dialog?.window?.let {window ->
            window.requestFeature(Window.FEATURE_NO_TITLE)
            window.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        }
        return inflater.inflate(R.layout.fragment_user_edit_about, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        configureView()
    }

    override fun dismiss() {
        userEditFragment?.scrollToOriginalPosition()
        super.dismiss()
    }

    // ----------------------------------
    // BUTTONS HANDLING
    // ----------------------------------
    fun onCloseClicked() {
        dismiss()
    }

    private fun onSaveClicked() {
        user_edit_about?.text.toString().let { about ->
            if (about.length > ABOUT_MAX_CHAR_COUNT) {
                Toast.makeText(context, R.string.user_edit_about_error, Toast.LENGTH_SHORT).show()
            } else {
                user?.let {u ->
                    u.about = about.trim { it <= ' ' }
                    userEditFragment?.initUserData() ?: run {
                        (parentFragmentManager.findFragmentByTag(UserFragment.TAG) as UserFragment?)?.saveAccount(u)
                    }
                }
                dismiss()
            }
        }
    }

    // ----------------------------------
    // PRIVATE METHODS
    // ----------------------------------
    private fun configureView() {
        userEditFragment = parentFragmentManager.findFragmentByTag(UserEditFragment.TAG) as UserEditFragment?
        if (userEditFragment == null) {
            //Try to see if we can find the User View fragment
            (parentFragmentManager.findFragmentByTag(UserFragment.TAG) as UserFragment?)?.let {userFragment ->
                user = userFragment.editedUser
            }
        } else {
            user = userEditFragment?.presenter?.editedUser
        }
        user_edit_about_close_button?.setOnClickListener { onCloseClicked()}
        user_edit_about_save_button?.setOnClickListener {onSaveClicked()}
        user?.let {u ->
            user_edit_about?.let {
                it.setText(u.about)
                it.setSelection(it.length())
                it.addTextChangedListener(object : TextWatcher {
                    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
                    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
                    override fun afterTextChanged(s: Editable) {
                        val charCountString = context!!.getString(R.string.entourage_create_title_char_count_format, s.length, ABOUT_MAX_CHAR_COUNT)
                        user_edit_about_char_count?.text = charCountString
                        if (s.length > ABOUT_MAX_CHAR_COUNT) {
                            user_edit_about_char_count?.setTextColor(ResourcesCompat.getColor(resources, R.color.entourage_error, null))
                        } else {
                            user_edit_about_char_count?.setTextColor(ResourcesCompat.getColor(resources, R.color.entourage_ok, null))
                        }
                    }
                })
                val charCountString = requireContext().getString(R.string.entourage_create_title_char_count_format, it.length(), ABOUT_MAX_CHAR_COUNT)
                user_edit_about_char_count?.text = charCountString
            }
        }
    }

    companion object {
        // ----------------------------------
        // CONSTANTS
        // ----------------------------------
        val TAG = UserEditAboutFragment::class.java.simpleName
        private const val ABOUT_MAX_CHAR_COUNT = 200
    }
}